[ADD] bonuses technicians

sbuhl opened the pull request:
The bonuses for technicians are calculated as follows:
For each sale, a quotation is validated and an order is generated. On this order, lines related to labor (main d'oeuvre) are added.
These labor lines have the following characteristics:
- They are products of the service type.
- The service_tracking field of these products is set to task_global_project | task_in project | project_only.
- The quantity of those product should be limited to 1 unit.
- The amount (HTVA), is equal to the labour_amount (see below)
Note: An order can contain multiple labor lines.

When the order is validated and the task(s) is/are created, one or more technicians will be assigned to the task(s).  Technicians must record the number of hours spent on the task on timesheets.
We need to retrieve the total number of hours worked (total_hours) as well as the number of hours per technician (tech1_hours, tech2_hours, etc.) in order to calculate the percentage of work per technician (tech1_percentage, tech2_percentage, etc.).

~The commission percentage granted to the technician will depend on what is recorded on the contract of the employee.
(If needed, later we could add another % on the product).~
Cost Recovery Rate Technicians (cost_recovery_rate_technicians). 
Note: previously, I wanted to add this on the contract of the technicians, but after reconsidering, I think it's better to add it on the product. As we want to keep what we pay under control.

Based on this percentage, the calculation will be as follows: The total cost recovery to give to technicians is equal to the labor amount from the SO times the cost recovery rate set on the product.
total_cost_recovery_technicans = cost_recovery_rate_technicians * labor_amount
Then this amount has to be distributed among technicians, based on their timesheet.
if technician A worked 5 hours, technician B worked 3 hours and technicians C worked 2 hours. The total cost should be slipped among them proportionnaly
technicianA_ cost_recovery_amount = total_cost_recovery_technicans / total_hours_worked * technicanA_total_hours worked
technicianB_ cost_recovery_amount = total_cost_recovery_technicans / total_hours_worked * technicanB_total_hours worked
technicianC_ cost_recovery_amount = total_cost_recovery_technicans / total_hours_worked * technicanC_total_hours worked
So each labour line has one task, and each task has at least one commission.

However, the bonus will only be "given" to the technician when the following conditions are met:

- The SO must be fully invoiced.
- All invoices must be fully paid.
- All products on the SO must be delivered.
- All labor services must be in the "Done" stage (thus, on the SO, the delivered quantity matches the ordered and invoiced quantity).

If these conditions are met (when the invoice is paid, or when the final products are delivered, or when the tasks are completed, etc.), a line must be added to the technician's vendor bill for receiving the bonus.
If there is no vendor bill yet, it needs to be created and addressed to the technician. On the line, a specific product should be used (which can be defined in the settings).
If there is an existing commission, still in draft, check if it's for the same month, else create a new bill.
One can't have commission from hibou and from labour, but if it happens it's fine, whatever if hibou vendor bill & labour vendor bill are different vendor bill.
A stat button on the invoice and on the SO should display the number of commissions related to this SO.
A stat button on the task should display the number of commissions related to this work.

In the settings, add a "Bonus" section.
Add a subsection for "Technician Bonuses," including the option to choose the product and the journal where the vendor bill will be located.

In the views, we should be able to see the technician bonuses from both the Field Services and Sales applications.

On the Invoice, there should be a "Calculate Bonus" button that will trigger the computation of the bonus.

sbuhl committed:
Initial commit

shortcut1337 committed:
m

shortcut1337 committed:
m

shortcut1337 committed:
m

shortcut1337 committed:
dont fail if no timesheet

shortcut1337 committed:
prevent bonus for SO before 1er juin

shortcut1337 committed:
cancel client invoice = cancel vendor bonus

sbuhl commented on models/bonus.py:
```diff
@@ -0,0 +1,170 @@
+        if any(line for line in order.order_line if line.product_uom_qty != line.qty_invoiced):
+            # Not all products have been invoiced
+            return
+        if any(line for line in order.order_line if line.product_uom_qty != line.qty_delivered):
```
@shortcut1337  Est-ce que tu peux regarder ici, uniquement pour les produits stockables ou consommables.

shortcut1337 committed:
few fixes
few more tests

no imp or behavior change

shortcut1337 committed:
handle

> cas 2:
> Le Bonus est créé
> On délie le paiement, on annule la facture (même si théoriquement, on ne peut pas faire ça comme ça)
> Il faut que le bonus s'annule.

Technically, any flow making an invoice go from "posted" to draft or
cancel will revert the bonuses

shortcut1337 committed:
fix bug from previous commit
factorize code
improve code
write tests for previous commit(s)

shortcut1337 committed:
IMP multi currencies
Add bonuses action stat btn from invoice
more tests

shortcut1337 committed:
regenerate action on SO

sbuhl commented on views/sale_order_views.xml:
```diff
@@ -0,0 +1,30 @@
+    </record>
+
+    <record id="action_invoice_order_regenerate_gse_bonuses" model="ir.actions.server">
+        <field name="name">Regenerate labor bonuses</field>
```
"Regenerate Transport Expenses"

sbuhl commented on views/sale_order_views.xml:
```diff
@@ -0,0 +1,30 @@
+        <field name="groups_id" eval="[(4, ref('account.group_account_manager'))]"/>
+        <field name="binding_model_id" ref="model_sale_order" />
+        <field name="state">code</field>
+        <field name="binding_view_types">form</field>
```
Could you also include it in the tree view?

sbuhl commented on views/sale_order_views.xml:
```diff
@@ -0,0 +1,30 @@
+            <xpath expr="//button[@name='action_view_invoice']" position="before">
+                <button type="object" name="action_view_bonuses" class="oe_stat_button" icon="fa-gift"
+                        attrs="{'invisible': [('bonuses_count', '=', 0)]}" groups="account.group_account_manager">
+                    <field name="bonuses_count" widget="statinfo" string="Bonuses"/>
```
string="Transport Expenses"

sbuhl commented on views/res_config_settings_views.xml:
```diff
@@ -0,0 +1,39 @@
+        <field name="inherit_id" ref="base.res_config_settings_view_form"/>
+        <field name="arch" type="xml">
+            <div name="performance" position="after">
+                <h2>Bonuses</h2>
```
Reword: Expenditures

sbuhl commented on views/res_config_settings_views.xml:
```diff
@@ -0,0 +1,39 @@
+                    <div class="col-xs-12 col-md-6 o_setting_box">
+                        <div class="o_setting_left_pane"/>
+                        <div class="o_setting_right_pane">
+                            <span class ="o_form_label">Technician Bonuses</span>
```
Reword: "Technician Transport Expenses"

sbuhl commented on views/res_config_settings_views.xml:
```diff
@@ -0,0 +1,39 @@
+                                    <field name="bonus_journal_id" domain="[('company_id', '=', company_id)]"/>
+                                </div>
+                                <div class="row">
+                                    <label string="Bonus Product" for="bonus_product_id" class="col-md-3 o_light_label"/>
```
string="Expense Product"

sbuhl commented on views/res_config_settings_views.xml:
```diff
@@ -0,0 +1,39 @@
+                                    <field name="bonus_product_id"/>
+                                </div>
+                                <div class="row">
+                                    <label string="Product Default Bonus Rate" for="bonus_rate" class="col-md-3 o_light_label"/>
```
string="Default Technician Expense Rate"

sbuhl commented on views/product_views.xml:
```diff
@@ -0,0 +1,15 @@
+        <field name="inherit_id" ref="sale.product_template_form_view"/>
+        <field name="arch" type="xml">
+            <field name="service_tracking" position="after">
+                <field name="bonus_rate" attrs="{'invisible': [('type', '!=', 'service')]}"/>
```
@shortcut1337  ceci ne permet-il pas de spécifier un rate spécifique par produit? 
Est-ce que cela ne résoud pas le "truc chiant" https://discord.com/channels/@me/695234725605802044/1177182352364011570 ?

Reword: string="Default Technician Expense Rate"

shortcut1337 replied:
C'est pas si simple, car `generate_bonuses()` ne donne un bonus que:
1. Si le produit est un produit de type service et qui génère une tâche (`service_tracking` field)
2. Et le bonus est calculé en fonction des timesheets liées à la tâche liée à la sale order line liée au produit

Ce qui était la spec initiale.
C'est pas très dur d'autoriser certains autres produits à générer un bonus, faut juste coder une condition spécial dans generate_bonus pour include un autre moyen d'ouvrir le droit à un bonus, et de coder la facon de générer le bonus pour ces types de produit.

Donc c'est un peu plus que juste mettre une valeur dans ce field

shortcut1337 replied:
En plus de tout ça, tout le modeling (fields) du `gse.bonus`tourne autour d'une timesheet.
Un bonus est otujours lié à une timesheet, c'est grace à la timesheet qu'on retrouve (computed fields):
- la SO line et donc la SO
- l'employee id
- la company et donc la currency

C'est en fait la partie la plus ennuyante mais rien d'impossible, je suis content d'avoir écrit des tests car ça peut du coup casser les bonus de timesheet si je change le modeling

shortcut1337 replied:
Voila c'est implémenté, je te laisserai testé, cf mon msg sur discord

sbuhl commented on views/bonus_views.xml:
```diff
@@ -0,0 +1,120 @@
+
+    <!-- Menu Items in Sales (sale_management) and Field Services (industry_fsm) -->
+    <menuitem id="menu_bonus_sale"
+        name="Bonuses"
```
Reword: "Technicians"

sbuhl commented on views/bonus_views.xml:
```diff
@@ -0,0 +1,120 @@
+        action="gse_bonuses.action_gse_bonus"
+        />
+    <menuitem id="menu_bonus_fsm"
+        name="Bonuses"
```
Reword: "Technicians"

sbuhl commented on security/security.xml:
```diff
@@ -0,0 +1,18 @@
+
+    <!-- Bonus Security Rules -->
+    <record id="gse_bonus_rule" model="ir.rule">
+        <field name="name">Bonus User</field>
```
name: Technician Transport Expenses User

sbuhl commented on security/security.xml:
```diff
@@ -0,0 +1,18 @@
+        <field name="groups" eval="[(4, ref('base.group_user'))]"/>
+    </record>
+    <record id="gse_bonus_rule_manager" model="ir.rule">
+        <field name="name">Bonus Manager</field>
```
name: Technician Transport Expenses Manager

sbuhl commented on models/bonus.py:
```diff
@@ -0,0 +1,212 @@
+        # - be fully paid
+        # - be fully delivered
+        # - have its related labor service tasks done (related SOL will be
+        #   marked as delivered, so it should be covered by previous point)
```
Can we add 2 fields and 2 conditions that have just emerged:
1. On an employee's contract, add a boolean: Allow Technician Transport Expenses.
2. On the task, add a boolean: Exclude from Technician Transport Expenses.
If the first is false, do not compute the bonus
If the second is true, do not compute the bonus.

sbuhl replied:
![Screenshot 2023-12-21 at 14 01 51](https://github.com/sbuhl/gse_bonuses/assets/505562/8d3e6fe6-d3d2-4921-ab02-9852b5a79aec)
![Screenshot 2023-12-21 at 14 04 01](https://github.com/sbuhl/gse_bonuses/assets/505562/8095a8e7-7891-4c29-98a6-eaff9937518e)

shortcut1337 replied:
Done

sbuhl reviewed (commented):
Hello @shortcut1337 
Thank you for the commits. 
I've 2 main requests before merge.
1. Few wording (we don't want to see "bonuses" in the interface but "transport expense"
2. Add 2 new conditions before creating a bonus. 

Do you think it's feasible in a relatively short timeframe, or should we go ahead and merge as is and make the changes afterwards?

shortcut1337 committed:
wording change

shortcut1337 committed:
truc chiant

shortcut1337 committed:
- new 2 conditions:
  - contract allow bonus
  - task allow bonus

- also fix warning in console because of manifest description

sbuhl committed:
Change Bonus computation to use subtotal

shortcut1337 committed:
- add hr contract to employees with `allow_transport_expenses` set to
  True, otherwise tests not working anymore (no bonus generated)
- add tax test for new commit about subtotal instead of total
  (without this tax_id set, the test fail if we revert the subtotal
   change commit)

shortcut1337 committed:
add test for new field `allow_transport_expenses`

shortcut1337 committed:
add test for new task field disallow_transport_expenses

shortcut1337 committed:
- fix singleton error when there is a SOL which is a note or a section
- + cover it by a test

shortcut1337 committed:
- add reversed invoice check
- quick bugfix which prevents to revert invoice sometimes

sbuhl committed:
[ADD] Add order date on technician view

sbuhl committed:
[FIX] Update Label: 'Bonus' to 'Technician Transport Expenses'

shortcut1337 commented:
Done at https://github.com/GoShop-Energy/technicians/pull/1
